libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -597,6 +601,186 @@ class plugin final : public virtual operator_plugin<deduplicate_operator> {
 
 } // namespace
 
+namespace {
+
+struct configuration2 {
+  ast::expression key_ = {};
+  std::optional<located<int64_t>> distance_ = {};
+  std::optional<located<int64_t>> limit_ = {};
+  std::optional<located<duration>> create_timeout_ = {};
+  std::optional<located<duration>> write_timeout_ = {};
+  std::optional<located<duration>> read_timeout_ = {};
+
+  friend auto inspect(auto& f, configuration2& x) -> bool {
+    return f.object(x).fields(f.field("key", x.key_),
+                              f.field("distance", x.distance_),
+                              f.field("limit", x.limit_),
+                              f.field("create_timeout", x.create_timeout_),
+                              f.field("write_timeout", x.write_timeout_),
+                              f.field("read_timeout", x.read_timeout_));
+  }
+};
+
+struct state2 {
+  int64_t count = {};
+  int64_t last_row = {};
+  std::chrono::steady_clock::time_point created_at = {};
+  std::chrono::steady_clock::time_point written_at = {};
+  std::chrono::steady_clock::time_point read_at = {};
+
+  void reset(int64_t current_row, std::chrono::steady_clock::time_point now) {
+    count = 1;
+    last_row = current_row;
+    created_at = now;
+    written_at = now;
+    read_at = now;
+  }
+};
+
+class deduplicate_operator2 final
+  : public crtp_operator<deduplicate_operator2> {
+public:
+  deduplicate_operator2() = default;
+
+  explicit deduplicate_operator2(configuration2 cfg) : cfg_(std::move(cfg)) {
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    tsl::robin_map<data, state2> state = {};
```

Comment:
The old deduplicate operator had an internal loop that cleaned up stale data with an expired distance or timeout. I still need to re-introduce that.